HIP/hip-844.md

```diff
@@ -0,0 +1,88 @@
+---
+hip: 844
+title: Handling and externalisation improvements for account nonce updates
+author: Stoyan Panayotov <stoyan.panayotov@limechain.tech>
+working-group: Nana Essilfie-Conduah <nana@swirldslabs.com>, Danno Ferrin <danno.ferrin@swirldslabs.com>, Steven Sheehy <steven.sheehy@swirldslabs.com>
+type: Standards Track
+needs-council-approval: Yes
+category: Service
+status: Review
+created: 2023-12-12
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/845
+---
+
+## Abstract
+
+Introduce rules for when to update an Ethereum transaction signer's nonce and a new field for externalising the nonce update in the record stream. This will help resolve issues where the Consensus nodes and the Mirror nodes are out of sync in regards to what the current value of an account's nonce is.
+
+## Motivation
+
+Unclear rules for when the nonce of an EthereumTransaction signer account should be updated result in discrepancies between the Consensus nodes' and Mirror nodes' states. This leads to users experiencing issues when reading the account nonce from the Mirror node and trying to use it in a transaction sent to the Consensus node.
+
+## Rationale
+
+The nonce of a signer of an Ethereum transaction should be increased for all transactions that are executed in the EVM.
+
+Transactions that have reached consensus and are handled by the node but have failed additional validations performed by the node prior to being executed in the EVM should have no effect on the Ethereum transaction signer nonce.
+
+A new protobuf field in the record stream - signerNonce - should be populated by the node with the current value of the signer account nonce after executing the transaction. This value should be used by the Mirror nodes when importing record stream files.
+
+## User stories
+
+1. As a Mirror node operator, I would like to receive information about incremented account nonces from the record stream produced by the consensus nodes.
+2. As a web3 user of the network, I would like to be able to more easily submit transactions without having to change the transaction nonce manualy.
+
+## Specification
+
+### Protobuf update
+
+Add a new protobuf field to contract_call_local.proto:
+
+```
+message ContractFunctionResult {
+    
+    [...]
+
+    /**
+     * If not null this field specifies what the value of the signer account nonce is post transaction execution. 
+     * For transactions that don't update the signer nonce (like HAPI ContractCall and ContractCreate transactions) this field should be null.
+     */
+    Int64Value signer_nonce = 15;
+}
+```
+
+### Services update
+
+1. Move the signer nonce update to be right before this code block that starts the transaction execution in the EVM:
```

Comment:
Instead of requiring a single line prior to this that could be refactored accidentally should we extract it to a single method that increments and initiates process e.g. `nonceIncrementAndEvmProcess()`

Reply:
Sounds like a good idea. But actually, I haven't analysed the modularised codebase implementation of the transaction processor and can't provide exact steps for the implementation fix. Should we leave this for the design doc/work package definition?